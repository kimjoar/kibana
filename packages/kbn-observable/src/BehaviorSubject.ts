import { Observer, Subscription } from './Observable';
import { Subject } from './Subject';
import { ObjectUnsubscribedError } from './errors';

export class BehaviorSubject<T> extends Subject<T> {
  constructor(private _value: T) {
    super();
  }

  protected _subscribe(observer: Observer<T>): Subscription {
    const subscription = super._subscribe(observer);

    if (subscription && !subscription.closed) {
      observer.next(this._value);
    }

    return subscription;
  }

  getValue() {
    if (this.hasError) {
      throw this.thrownError;
    } else if (this.closed) {
      throw new ObjectUnsubscribedError();
    }

    return this._value;
  }

  get value(): T {
    return this.getValue();
  }

  next(value: T): void {
    this._value = value;
    super.next(value);
  }
}
