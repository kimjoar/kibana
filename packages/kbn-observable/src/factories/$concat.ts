import { Observable, Subscription } from '../Observable';

/**
 * Creates an observable that combines all observables passed as arguments into
 * a single output observable by subscribing to them in series.
 *
 * @param {Observable...}
 * @return {Observable}
 */
export function $concat<T>(...observables: Observable<T>[]) {
  return new Observable(observer => {
    let subscription: Subscription;

    function subscribe(i: number) {
      if (observer.closed) {
        return;
      }

      if (i >= observables.length) {
        observer.complete();
      }

      subscription = observables[i].subscribe({
        next(value) {
          observer.next(value);
        },
        error(error) {
          observer.error(error);
        },
        complete() {
          subscribe(i + 1);
        }
      });
    }

    subscribe(0);

    return function() {
      if (subscription) {
        subscription.unsubscribe();
      }
    };
  });
}
