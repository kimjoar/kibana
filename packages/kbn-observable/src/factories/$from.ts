import { tryCoerce } from '../lib';
import { Observable, ObservableInput } from '../Observable';

/**
 *  Convert an input into an Observable based on the following rules:
 *
 *  if `input` is:
 *   - falsy : throw an error
 *   - not an object: throw an error
 *   - a Native Observable: return the value
 *   - `input[Symbol.observable]` is a function, see $fromSubscribable
 *   - `input[Symbol.iterator]` is a function, see $fromIterable
 *   - if `input` is a promise, see $fromPromise
 *
 *  @param {Observable|Subscribable|Iterator|Promise}
 *  @return {Observable}
 */
export function $from<T>(input: ObservableInput<T>): Observable<T> {
  const coerced = tryCoerce(input);

  if (coerced === undefined) {
    throw new TypeError(`${input} can't be coerced to an observable`);
  }

  return coerced;
}
