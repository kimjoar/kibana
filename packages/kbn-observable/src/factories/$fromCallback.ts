import { Observable } from '../Observable';

/**
 * Creates an observable that calls the specified function with no arguments
 * when it is subscribed. The observerable will behave differently based on the
 * return value of the factory:
 *
 * - return `undefined`: observable will immediately complete
 * - otherwise: observable will emit the value and then complete
 *
 * @param {Function}
 * @returns {Observable}
 */
export function $fromCallback<T>(factory: () => T): Observable<T> {
  return new Observable(observer => {
    const result = factory();

    // empty stream
    if (result === undefined) {
      observer.complete();
      return;
    }

    // result is just a value
    observer.next(result as T);
    observer.complete();
  });
}
