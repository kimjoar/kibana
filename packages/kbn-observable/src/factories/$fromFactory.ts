import { Observable, ObservableInput } from '../Observable';
import { tryCoerce } from '../lib';

/**
 * Creates an observable that calls a factory function when it is subscribed.
 * The observerable will behave differently based on the return value of the
 * factory,
 *
 * - return `undefined`:
 *   observable will immediately complete
 * - return a value compatible with `$coerce()`:
 *   observerable will mirror the coerced value
 * - anything else:
 *   observable will emit the value and then complete
 *
 * @param {Function}
 * @returns {Observable}
 */
export function $fromFactory<T>(
  factory: () => T | ObservableInput<T>
): Observable<T> {
  return new Observable(observer => {
    const result = factory();

    // empty stream
    if (result === undefined) {
      observer.complete();
      return;
    }

    // TODO decide if this is needed or not

    // try to coerce the result into an observable
    const coerced = tryCoerce<T>(result);

    if (coerced !== undefined) {
      return coerced.subscribe(observer);
    }

    // result is just a value
    observer.next(result as T);
    observer.complete();
  });
}
