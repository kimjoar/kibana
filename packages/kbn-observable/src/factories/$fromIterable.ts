import { Observable } from '../Observable';

/**
 * Creates an Observable from an iterable value, like Arrays or Generators.
 *
 * @param {Iterable<T>}
 * @return {Observable<T>}
 */
export function $fromIterable<T>(iterable: Iterable<T>): Observable<T> {
  return new Observable(observer => {
    for (const value of iterable) {
      observer.next(value);

      if (observer.closed) {
        return;
      }
    }

    observer.complete();
  });
}
