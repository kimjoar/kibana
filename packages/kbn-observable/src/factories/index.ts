export { $from } from './$from';
export { $combineLatest } from './$combineLatest';
export { $concat } from './$concat';
export { $fromFactory } from './$fromFactory';
export { $fromIterable } from './$fromIterable';
export { $fromPromise } from './$fromPromise';
export { $fromSubscribable } from './$fromSubscribable';
export { $of } from './$of';
