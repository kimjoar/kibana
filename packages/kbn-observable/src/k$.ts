import { pipeFromArray } from './lib';
import { UnaryFunction } from './interfaces';

export function k$<T, A>(source: T, op1: UnaryFunction<T, A>): A;
export function k$<T, A, B>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>
): B;
export function k$<T, A, B, C>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>
): C;
export function k$<T, A, B, C, D>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>
): D;
export function k$<T, A, B, C, D, E>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>,
  op5: UnaryFunction<D, E>
): E;
export function k$<T, A, B, C, D, E, F>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>,
  op5: UnaryFunction<D, E>,
  op6: UnaryFunction<E, F>
): F;
export function k$<T, A, B, C, D, E, F, G>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>,
  op5: UnaryFunction<D, E>,
  op6: UnaryFunction<E, F>,
  op7: UnaryFunction<F, G>
): G;
export function k$<T, A, B, C, D, E, F, G, H>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>,
  op5: UnaryFunction<D, E>,
  op6: UnaryFunction<E, F>,
  op7: UnaryFunction<F, G>,
  op8: UnaryFunction<G, H>
): H;
export function k$<T, A, B, C, D, E, F, G, H, I>(
  source: T,
  op1: UnaryFunction<T, A>,
  op2: UnaryFunction<A, B>,
  op3: UnaryFunction<B, C>,
  op4: UnaryFunction<C, D>,
  op5: UnaryFunction<D, E>,
  op6: UnaryFunction<E, F>,
  op7: UnaryFunction<F, G>,
  op8: UnaryFunction<G, H>,
  op9: UnaryFunction<H, I>
): I;

export function k$<T, R>(source: T, ...operations: UnaryFunction<T, R>[]): R {
  return pipeFromArray(operations)(source);
}
