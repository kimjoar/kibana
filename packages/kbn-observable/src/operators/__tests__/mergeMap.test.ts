import { Observable } from '../../Observable';
import { k$ } from '../../k$';
import { BehaviorSubject } from '../../BehaviorSubject';
import { mergeMap, delay } from '../';
import { $of, $concat, $from } from '../../factories';

it('should map values to constant resolved promises and merge', done => {
  expect.assertions(1);

  const source = Observable.from([4, 3, 2, 1]);
  const results: any[] = [];

  k$(source, mergeMap(() => Promise.resolve(42))).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      expect(results).toEqual([42, 42, 42, 42]);
      done();
    }
  });
});

it('should map values to constant rejected promises and merge', done => {
  expect.assertions(1);

  const source = Observable.from([4, 3, 2, 1]);

  k$(source, mergeMap(() => Promise.reject(42))).subscribe({
    next(val) {
      done(new Error(`next handled unexpectedly called with value: ${val}`));
    },
    error(err) {
      expect(err).toBe(42);
      done();
    },
    complete() {
      done(new Error(`complete handled unexpectedly called`));
    }
  });
});

it('should map values to resolved promises and merge', done => {
  expect.assertions(1);

  const source = Observable.from([4, 3, 2, 1]);
  const project = (value: number, index: number) =>
    Promise.resolve(value + index);

  const results: number[] = [];
  k$(source, mergeMap(project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      expect(results).toEqual([4, 4, 4, 4]);
      done();
    }
  });
});

it('should map values to rejected promises and merge', done => {
  expect.assertions(1);

  const source = Observable.from([4, 3, 2, 1]);
  const project = (value: number, index: number) =>
    Promise.reject(`${value} / ${index}`);

  k$(source, mergeMap(project)).subscribe({
    next(val) {
      done(new Error(`next handled unexpectedly called with value: ${val}`));
    },
    error(err) {
      expect(err).toBe('4 / 0');
      done();
    },
    complete() {
      done(new Error(`complete handled unexpectedly called`));
    }
  });
});

it('should mergeMap values to resolved promises with resultSelector', done => {
  expect.assertions(2);

  const source = Observable.from([4, 3, 2, 1]);
  const resultSelector = jest.fn(() => 8);
  const project = (value: number, index: number) =>
    Promise.resolve([value, index]);

  const results: any = [];

  k$(source, mergeMap(project, resultSelector)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      expect(results).toEqual([8, 8, 8, 8]);
      expect(resultSelector.mock.calls).toMatchSnapshot();
      done();
    }
  });
});

it('should mergeMap values to rejected promises with resultSelector', done => {
  expect.assertions(1);

  const source = Observable.from([4, 3, 2, 1]);
  const project = (value: number, index: number) =>
    Promise.reject(`${value} / ${index}`);
  const resultSelector = () => {
    throw 'this should not be called';
  };

  k$(source, mergeMap(project, resultSelector)).subscribe({
    next(val) {
      done(new Error(`next handled unexpectedly called with value: ${val}`));
    },
    error(err) {
      expect(err).toBe('4 / 0');
      done();
    },
    complete() {
      done(new Error(`complete handled unexpectedly called`));
    }
  });
});

it('should mergeMap many outer values to many inner values', done => {
  expect.assertions(1);

  const source = Observable.from([1, 2, 3, 4]);
  const project = (value: number, index: number) =>
    $concat(
      k$($of(`${value}-a`), delay(5)),
      k$($of(`${value}-b`), delay(10)),
      k$($of(`${value}-c`), delay(15))
    );

  const results: any[] = [];
  k$(source, mergeMap(project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      const expected = [
        '1-a',
        '2-a',
        '3-a',
        '4-a',
        '1-b',
        '2-b',
        '3-b',
        '4-b',
        '1-c',
        '2-c',
        '3-c',
        '4-c'
      ];
      expect(results).toEqual(expected);
      done();
    }
  });
});

it('should mergeMap many outer values to many inner values, early complete', done => {
  expect.assertions(1);

  const source = new BehaviorSubject(1);
  const project = (value: number, index: number) =>
    $concat(
      k$($of(`${value}-a`), delay(5)),
      k$($of(`${value}-b`), delay(10)),
      k$($of(`${value}-c`), delay(15))
    );

  const results: any[] = [];
  k$(source, mergeMap(project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      const expected = [
        '1-a',
        '2-a',
        '3-a',
        '1-b',
        '2-b',
        '3-b',
        '1-c',
        '2-c',
        '3-c'
      ];
      expect(results).toEqual(expected);
      done();
    }
  });

  source.next(2);
  source.next(3);
  source.complete();

  // This shouldn't end up in the results
  source.next(4);
});

it('should mergeMap many outer to many inner, and inner throws', done => {
  expect.assertions(1);

  const source = Observable.from([1, 2, 3, 4]);
  const project = (value: number, index: number) => {
    if (index > 1) {
      return new Observable(observer => {
        observer.error(new Error('fail'));
      });
    }
    return Observable.of(value);
  };

  const results: any[] = [];
  k$(source, mergeMap(project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      expect(results).toEqual([1, 2]);
      done();
    },
    complete() {
      done(new Error(`complete handled unexpectedly called`));
    }
  });
});

it('should mergeMap many outer to many inner, and outer throws', done => {
  expect.assertions(1);

  const source = new BehaviorSubject(1);
  const project = (value: number, index: number) =>
    $concat(
      k$($of(`${value}-a`), delay(5)),
      k$($of(`${value}-b`), delay(10)),
      k$($of(`${value}-c`), delay(15))
    );

  const results: any[] = [];
  k$(source, mergeMap(project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      expect(results).toEqual(['1-a', '2-a', '3-a', '1-b', '2-b', '3-b']);
      done();
    },
    complete() {
      done(new Error('complete handled unexpectedly called'));
    }
  });

  source.next(2);
  source.next(3);

  setTimeout(() => {
    source.error(new Error('outer fails'));
    // This shouldn't end up in the results
    source.next(4);
  }, 20);
});

it('should mergeMap many outer to an array for each value', done => {
  expect.assertions(1);

  const source = Observable.from([1, 2, 3]);
  const results: any[] = [];

  k$(source, mergeMap(() => $of('a', 'b', 'c'))).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      expect(results).toEqual(['a', 'b', 'c', 'a', 'b', 'c', 'a', 'b', 'c']);
      done();
    }
  });
});

it('should mergeMap many outer to inner arrays, using resultSelector', done => {
  expect.assertions(1);

  const source = Observable.from([1, 2, 3]);
  const results: any[] = [];

  const project = (num: number, str: string) => `${num}/${str}`;

  k$(source, mergeMap(() => $of('a', 'b', 'c'), project)).subscribe({
    next(x) {
      results.push(x);
    },
    error(err) {
      done(err);
    },
    complete() {
      expect(results).toEqual([
        '1/a',
        '1/b',
        '1/c',
        '2/a',
        '2/b',
        '2/c',
        '3/a',
        '3/b',
        '3/c'
      ]);
      done();
    }
  });
});
