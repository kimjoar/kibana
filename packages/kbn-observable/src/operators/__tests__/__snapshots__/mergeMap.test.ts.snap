// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should mergeMap values to resolved promises with resultSelector 1`] = `
Array [
  Array [
    4,
    Array [
      4,
      0,
    ],
    0,
    0,
  ],
  Array [
    3,
    Array [
      3,
      1,
    ],
    1,
    0,
  ],
  Array [
    2,
    Array [
      2,
      2,
    ],
    2,
    0,
  ],
  Array [
    1,
    Array [
      1,
      3,
    ],
    3,
    0,
  ],
]
`;
